### PR TITLE
[WIP] Asynchronous Network  

### PR DESCRIPTION
**Pull request by aarani**
I created this as a PoC it works but it's not great I want to propose a new branch for everyone to work on this as it's extremely necessary for having a functional telegram api 
PS: I use NetCoreServer library and it needs a .NET Framework 4.6 Update

### CODE REVIEW

**Comment by aarani** [comment]
@knocte Please check this commit and tell me your opinion

File: TLSharp.Core/Network/TcpTransport.cs [hunk 1]
```diff
@@ -10,103 +12,64 @@ namespace TLSharp.Core.Network
...
+        private readonly BlockingCollection<TcpMessage> ReceievedMessage = new BlockingCollection<TcpMessage>();
+        //private readonly NetworkStream stream;
         private int sendCounter = 0;
 
         public TcpTransport(string address, int port, TcpClientConnectionHandler handler = null)
         {
-            if (handler == null)
-            {
-                var ipAddress = IPAddress.Parse(address);
-                var endpoint = new IPEndPoint(ipAddress, port);
```
Comment (aarani) [inline TLSharp.Core/Network/TcpTransport.cs:22]: As far as i know this library does not handle handlers and proxy so we should work on an implementation later.

File: TLSharp.Core/Network/TcpTransport.cs [hunk 2]
```diff
@@ -10,103 +12,64 @@ namespace TLSharp.Core.Network
...
-
-            byte[] rv = new byte[packetLengthBytes.Length + seqBytes.Length + body.Length];
-
-            Buffer.BlockCopy(packetLengthBytes, 0, rv, 0, packetLengthBytes.Length);
-            Buffer.BlockCopy(seqBytes, 0, rv, packetLengthBytes.Length, seqBytes.Length);
-            Buffer.BlockCopy(body, 0, rv, packetLengthBytes.Length + seqBytes.Length, body.Length);
-            var crc32 = new Ionic.Crc.CRC32();
-            crc32.SlurpBlock(rv, 0, rv.Length);
-            var validChecksum = crc32.Crc32Result;
+            return ReceievedMessage.Take();
```
Comment (aarani) [inline TLSharp.Core/Network/TcpTransport.cs:52]: Old Code Compatibility Only We need to get rid of this

**Comment by knocte** [comment]
@CheshireCaat can you test this?

File: TLSharp.Core/Network/TcpTransport.cs [hunk 3]
```diff
@@ -10,103 +12,64 @@ namespace TLSharp.Core.Network
...
+        }
 
-                tcpClient = new TcpClient(ipAddress.AddressFamily);
-                tcpClient.Connect(endpoint);
-            }
-            else
-                tcpClient = handler(address, port);
+        public void ReceiveMessage(TcpMessage message)
+        {
+            ReceievedMessage.Add(message);
```
Comment (aarani) [inline TLSharp.Core/Network/TcpTransport.cs:34]: Old Code Compatibility. It's here to make sync receive function work.
In the final version we should read the message deserialize it find the type and call the handler for that type.
for example if we get a TLMessage we should search a dictionary<type,delegate> to find the handler function.

**Comment by knocte** [comment]
@aarani : to begin with, can you separate this into 2 PRs? another one which just upgrades the .NETFramework version

**Comment by aarani** [comment]
> @CheshireCaat can you test this?

I should say that i used a console application to test it because i couldn't update the test projects to .net framework 4.6

**Comment by aarani** [comment]
> @aarani : to begin with, can you separate this into 2 PRs? another one which just upgrades the .NETFramework version

Sure, I try doing it but test projects are pain to update

**Comment by knocte** [comment]
> because i couldn't update the test projects to .net framework 4.6

why you couldn't upgrade them?

**Comment by knocte** [comment]
> Sure, I try doing it but test projects are pain to update

why a pain? please share more information with us, we're also developers, not managers :)

**Comment by aarani** [comment]
> > because i couldn't update the test projects to .net framework 4.6
> 
> why you couldn't upgrade them?

Visual studio gave me a weird error when I tried doing it so I did not bother to figure out what's wrong. I try again.

Capture

**Comment by aarani** [comment]
> > Sure, I try doing it but test projects are pain to update
> 
> why a pain? please share more information with us, we're also developers, not managers :)

Sure, but I wanted to take a look at it myself first and then bother you guys with it 😄

**Comment by knocte** [comment]
> Visual studio gave me a weird error when I tried doing it so I did not bother to figure out what's wrong. I try again.

I think this happened me once, when I wrote "4.6.0" instead of "4.6" (the latter is the correct version number). Maybe this was the mistake?

If not, try simply restarting Visual Studio like this guy says here: https://stackoverflow.com/a/56597422/544947

**Comment by aarani** [comment]
> > Visual studio gave me a weird error when I tried doing it so I did not bother to figure out what's wrong. I try again.
> 
> I think this happened to me once, when I wrote "4.6.0" instead of "4.6" (the latter is the correct version number). Maybe this was the mistake?
> 
> If not, try simply restarting Visual Studio like this guy says here: https://stackoverflow.com/a/56597422/544947

wow restarting worked 😆 I will try running the tests too.

**Comment by aarani** [comment]
> @aarani : to begin with, can you separate this into 2 PRs? another one which just upgrades the .NETFramework version

#920

**Comment by aarani** [comment]
Sorry, I accidentally pushed my api_id and api_hash and had to force push it this pull_request is toast

**Pull request closed without merging**